File: hw/total/white_box/BST_pred/code/python/pred.py
from BinNode import BinNode


def pred(node):
    # find the pred of the node
    if node.lc is not None:
        node = node.lc
        while True:
            if node.rc is not None:
                node = node.rc
            else:
                break
        return node
    y = node.p
    x = node
    while True:
        if y is not None and x == y.lc:
            x = y
            y = y.p
        else:
            break
    return y

File: hw/total/white_box/BST_pred/code/java/pred.java
public class pred {
	public BinNode findPred(BinNode node) {
		if (node.has_lc()) {
			node = node.lc;
			while (true) {
				if (node.has_rc())
					node = node.rc;
				else
					break;
			}
			return node;
		}
		BinNode y = node.p;
		BinNode x = node;
		while (true) {
			if (y != null && x == y.lc) {
				x = y;
				y = y.p;
			}
			else
				break;
		}
		return y;
	}	
}
